### PR TITLE
Fixes #9: Add different sizes for different platforms

### LINKED ISSUE #9: Different sizes for different platform
Sasha supports only 1024x1024 original image. But some platforms require small sizes. For example, max size for watchOS complication icons is 203x203. It would be great to check max size for selected platform.

### PR DESCRIPTION
**Pull request by imvm**
These changes calculate the maximum pixel size for each platform and use that to determine the Keys.defaultSize property. So far the only difference is that if the user wants to use a smaller image to generate say a set of complication images, he can. It would be nice to also add a query where the user could check the max sizes per platform without having to necessarily generate images.

### CODE REVIEW

File: Sources/SashaCore/Services/IconService.swift [hunk 1]
```diff
@@ -24,7 +24,7 @@ final class IconService {
         static let complicationSetName = "Complication.complicationset"
         static let contentsName = "Contents.json"
         static let androidIconFolder = "Android"
-        static let defaultSize: Float = 1024
+        static var defaultSize: Float = 0
     }
```
Comment (artemnovichkov) [inline Sources/SashaCore/Services/IconService.swift:29]: I guess you can remove this variable and use only local variable

File: Sources/SashaCore/Services/IconService.swift [hunk 2]
```diff
@@ -77,7 +77,6 @@ final class IconService {
     /// - Parameter output: Output path for generated icons. Default value is nil.
     /// - Throws: `IconService.Error` errors.
     func generateWatchOSComplicationIcons(for imageURL: URL, output: String? = nil) throws {
-        let image = try self.image(for: imageURL)
         let idioms: [Icon.Idiom] = [.complicationCircular,
```
Comment (artemnovichkov) [inline Sources/SashaCore/Services/IconService.swift]: Why did you move this line below? It produces an errors before set creating

File: Sources/SashaCore/Services/IconService.swift [hunk 3]
```diff
@@ -142,22 +151,33 @@ final class IconService {
             let height = image.properties[Keys.height] as? Float else {
                 throw Error.sizeReadingFailed
         }
-        guard width == Keys.defaultSize, height == Keys.defaultSize else {
+        guard width >= Keys.defaultSize, height >= Keys.defaultSize else {
             throw Error.wrongSizeDetected
```
Comment (artemnovichkov) [inline Sources/SashaCore/Services/IconService.swift]: This logic will break check of non-square images

File: Sources/SashaCore/Services/IconService.swift [hunk 4]
```diff
@@ -142,22 +151,33 @@ final class IconService {
...
+    func maxSize(for iconSet: AppIconSet) -> Float {
+        let max = iconSet.icons.map { $0.size * $0.scale }.max()
+        return max ?? 0
+    }
+
+    func maxSize(for iconSet: [AndroidIcon]) -> Float {
+        let max = iconSet.map { $0.size }.max()
+        return max ?? 0
+    }
+
```
Comment (artemnovichkov) [inline Sources/SashaCore/Services/IconService.swift]: Please add `private` keyword for these functions

File: Sources/SashaCore/Services/IconService.swift [hunk 5]
```diff
@@ -142,22 +151,33 @@ final class IconService {
...
+        guard width >= Keys.defaultSize, height >= Keys.defaultSize else {
             throw Error.wrongSizeDetected
         }
         return image
     }
 
+    func maxSize(for iconSet: AppIconSet) -> Float {
+        let max = iconSet.icons.map { $0.size * $0.scale }.max()
+        return max ?? 0
+    }
```
Comment (artemnovichkov) [inline Sources/SashaCore/Services/IconService.swift]: What's about adding `maxSize` to `AppIconSet`?

**Review by artemnovichkov — CHANGES_REQUESTED** [review]

File: Sources/SashaCore/Services/IconService.swift (hunk 1 above)
Comment (imvm) [reply Sources/SashaCore/Services/IconService.swift:29]: Removed

**Comment by imvm** [comment]
After the review I thought about checking the scale factor on a per-icon basis, on-the-fly. This still allows for different sizes for different platforms, since the required image size isn't based on a default value but instead on the specific icon sizes. This method avoids the need of a global defaultSize variable or maxSize computed property for the AppIconSet class since we would only compare the icon being currently generated and the source image. I believe this also simplifies the code considering its current state, although later on a different strategy might be better suited. This is because the changes will be concentrated in the generateIcons function, instead of spread throughout all the platform specific functions. A potential drawback is that several icons may be generated before the program finds a incompatible icon. Performance impact should be minimal at most. In the future, if TVOS icons are supported, non-square images could be accepted by having height and width variables in the icons and checking the source image aspect ratio.

**Comment by artemnovichkov** [comment]
Hi, @imvm! Thanks for the changes! I'll review it asap.

File: Sources/SashaCore/Services/IconService.swift [hunk 6]
```diff
@@ -240,6 +231,33 @@ final class IconService {
...
         }
     }
+
+    /// Checks icon and image size compatibility and returns the scale factor between them
+    ///
+    /// - Parameters:
+    ///   - image: The original image.
+    ///   - icon: Object that represents icon information.
+    /// - Throws: `IconService.Error` errors.
+    func scaleIcon(_ icon: IconRepresentable, with image: CIImage) throws -> Float {
```
Comment (artemnovichkov) [inline Sources/SashaCore/Services/IconService.swift:241]: `IconRepresentable` is a custom type, so we can omit its name in the function name
```suggestion
    func scale(_ icon: IconRepresentable, with image: CIImage) throws -> Float {
```

**Review by artemnovichkov — CHANGES_REQUESTED** [review]

File: Sources/SashaCore/Services/IconService.swift [hunk 7]
```diff
@@ -240,6 +231,33 @@ final class IconService {
...
+        }
+
+        //check for smaller image than necessary
+        guard width >= icon.iconSize, height >= icon.iconSize else {
+            throw Error.wrongSizeDetected(actualWidth: width, actualHeight: height,
+                                          expectedWidth: icon.iconSize, expectedHeight: icon.iconSize)
+        }
+
+        //check for non-square image (needs to be adapted later for non-square Apple TV icons)
+        guard width == height else {
```
Comment (artemnovichkov) [inline Sources/SashaCore/Services/IconService.swift:254]: Both guards produce the same error. What's about using descriptive states? For example:
```swift
let greatOrEqualIconSize = width >= icon.iconSize, height >= icon.iconSize
let isSquared = width == height
guard greatOrEqualIconSize, isSquared else {
    throw Error.wrongSizeDetected(actualWidth: width, actualHeight: height, expectedWidth: icon.iconSize, expectedHeight: icon.iconSize)
}
```
What do you think?

**Review by artemnovichkov — CHANGES_REQUESTED** [review]
